<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN" "http://www.w3.org/TR/html4/loose.dtd">
<!-- NewPage -->
<html lang="en">
<head>
<!-- Generated by javadoc (1.8.0_91) on Mon Jul 10 19:09:19 EDT 2017 -->
<title>Uses of Class bioLockJ.Module</title>
<meta name="date" content="2017-07-10">
<link rel="stylesheet" type="text/css" href="../../stylesheet.css" title="Style">
<script type="text/javascript" src="../../script.js"></script>
</head>
<body>
<script type="text/javascript"><!--
    try {
        if (location.href.indexOf('is-external=true') == -1) {
            parent.document.title="Uses of Class bioLockJ.Module";
        }
    }
    catch(err) {
    }
//-->
</script>
<noscript>
<div>JavaScript is disabled on your browser.</div>
</noscript>
<!-- ========= START OF TOP NAVBAR ======= -->
<div class="topNav"><a name="navbar.top">
<!--   -->
</a>
<div class="skipNav"><a href="#skip.navbar.top" title="Skip navigation links">Skip navigation links</a></div>
<a name="navbar.top.firstrow">
<!--   -->
</a>
<ul class="navList" title="Navigation">
<li><a href="../../overview-summary.html">Overview</a></li>
<li><a href="../package-summary.html">Package</a></li>
<li><a href="../../bioLockJ/Module.html" title="class in bioLockJ">Class</a></li>
<li class="navBarCell1Rev">Use</li>
<li><a href="../package-tree.html">Tree</a></li>
<li><a href="../../deprecated-list.html">Deprecated</a></li>
<li><a href="../../index-files/index-1.html">Index</a></li>
<li><a href="../../help-doc.html">Help</a></li>
</ul>
</div>
<div class="subNav">
<ul class="navList">
<li>Prev</li>
<li>Next</li>
</ul>
<ul class="navList">
<li><a href="../../index.html?bioLockJ/class-use/Module.html" target="_top">Frames</a></li>
<li><a href="Module.html" target="_top">No&nbsp;Frames</a></li>
</ul>
<ul class="navList" id="allclasses_navbar_top">
<li><a href="../../allclasses-noframe.html">All&nbsp;Classes</a></li>
</ul>
<div>
<script type="text/javascript"><!--
  allClassesLink = document.getElementById("allclasses_navbar_top");
  if(window==top) {
    allClassesLink.style.display = "block";
  }
  else {
    allClassesLink.style.display = "none";
  }
  //-->
</script>
</div>
<a name="skip.navbar.top">
<!--   -->
</a></div>
<!-- ========= END OF TOP NAVBAR ========= -->
<div class="header">
<h2 title="Uses of Class bioLockJ.Module" class="title">Uses of Class<br>bioLockJ.Module</h2>
</div>
<div class="classUseContainer">
<ul class="blockList">
<li class="blockList">
<table class="useSummary" border="0" cellpadding="3" cellspacing="0" summary="Use table, listing packages, and an explanation">
<caption><span>Packages that use <a href="../../bioLockJ/Module.html" title="class in bioLockJ">Module</a></span><span class="tabEnd">&nbsp;</span></caption>
<tr>
<th class="colFirst" scope="col">Package</th>
<th class="colLast" scope="col">Description</th>
</tr>
<tbody>
<tr class="altColor">
<td class="colFirst"><a href="#bioLockJ.module.classifier">bioLockJ.module.classifier</a></td>
<td class="colLast">&nbsp;</td>
</tr>
<tr class="rowColor">
<td class="colFirst"><a href="#bioLockJ.module.classifier.r16s">bioLockJ.module.classifier.r16s</a></td>
<td class="colLast">&nbsp;</td>
</tr>
<tr class="altColor">
<td class="colFirst"><a href="#bioLockJ.module.classifier.r16s.qiime">bioLockJ.module.classifier.r16s.qiime</a></td>
<td class="colLast">&nbsp;</td>
</tr>
<tr class="rowColor">
<td class="colFirst"><a href="#bioLockJ.module.classifier.wgs">bioLockJ.module.classifier.wgs</a></td>
<td class="colLast">&nbsp;</td>
</tr>
<tr class="altColor">
<td class="colFirst"><a href="#bioLockJ.module.parser">bioLockJ.module.parser</a></td>
<td class="colLast">&nbsp;</td>
</tr>
<tr class="rowColor">
<td class="colFirst"><a href="#bioLockJ.module.parser.r16s">bioLockJ.module.parser.r16s</a></td>
<td class="colLast">&nbsp;</td>
</tr>
<tr class="altColor">
<td class="colFirst"><a href="#bioLockJ.module.parser.wgs">bioLockJ.module.parser.wgs</a></td>
<td class="colLast">&nbsp;</td>
</tr>
<tr class="rowColor">
<td class="colFirst"><a href="#bioLockJ.module.postProcessor">bioLockJ.module.postProcessor</a></td>
<td class="colLast">&nbsp;</td>
</tr>
<tr class="altColor">
<td class="colFirst"><a href="#bioLockJ.module.preProcessor">bioLockJ.module.preProcessor</a></td>
<td class="colLast">&nbsp;</td>
</tr>
<tr class="rowColor">
<td class="colFirst"><a href="#bioLockJ.util">bioLockJ.util</a></td>
<td class="colLast">&nbsp;</td>
</tr>
</tbody>
</table>
</li>
<li class="blockList">
<ul class="blockList">
<li class="blockList"><a name="bioLockJ.module.classifier">
<!--   -->
</a>
<h3>Uses of <a href="../../bioLockJ/Module.html" title="class in bioLockJ">Module</a> in <a href="../../bioLockJ/module/classifier/package-summary.html">bioLockJ.module.classifier</a></h3>
<table class="useSummary" border="0" cellpadding="3" cellspacing="0" summary="Use table, listing subclasses, and an explanation">
<caption><span>Subclasses of <a href="../../bioLockJ/Module.html" title="class in bioLockJ">Module</a> in <a href="../../bioLockJ/module/classifier/package-summary.html">bioLockJ.module.classifier</a></span><span class="tabEnd">&nbsp;</span></caption>
<tr>
<th class="colFirst" scope="col">Modifier and Type</th>
<th class="colLast" scope="col">Class and Description</th>
</tr>
<tbody>
<tr class="altColor">
<td class="colFirst"><code>class&nbsp;</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../bioLockJ/module/classifier/ClassifierModule.html" title="class in bioLockJ.module.classifier">ClassifierModule</a></span></code>
<div class="block">This is the Classifier superclass used by all WGS & 16S classifiers.</div>
</td>
</tr>
</tbody>
</table>
</li>
<li class="blockList"><a name="bioLockJ.module.classifier.r16s">
<!--   -->
</a>
<h3>Uses of <a href="../../bioLockJ/Module.html" title="class in bioLockJ">Module</a> in <a href="../../bioLockJ/module/classifier/r16s/package-summary.html">bioLockJ.module.classifier.r16s</a></h3>
<table class="useSummary" border="0" cellpadding="3" cellspacing="0" summary="Use table, listing subclasses, and an explanation">
<caption><span>Subclasses of <a href="../../bioLockJ/Module.html" title="class in bioLockJ">Module</a> in <a href="../../bioLockJ/module/classifier/r16s/package-summary.html">bioLockJ.module.classifier.r16s</a></span><span class="tabEnd">&nbsp;</span></caption>
<tr>
<th class="colFirst" scope="col">Modifier and Type</th>
<th class="colLast" scope="col">Class and Description</th>
</tr>
<tbody>
<tr class="altColor">
<td class="colFirst"><code>class&nbsp;</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../bioLockJ/module/classifier/r16s/QiimeClassifier.html" title="class in bioLockJ.module.classifier.r16s">QiimeClassifier</a></span></code>
<div class="block">QiimeClassifier is a superclass to several classes in the qiime package so hold shared methods
 for these classes.</div>
</td>
</tr>
<tr class="rowColor">
<td class="colFirst"><code>class&nbsp;</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../bioLockJ/module/classifier/r16s/RdpClassifier.html" title="class in bioLockJ.module.classifier.r16s">RdpClassifier</a></span></code>
<div class="block">RdpClassifier is used to build RDP classifier bash scripts</div>
</td>
</tr>
</tbody>
</table>
</li>
<li class="blockList"><a name="bioLockJ.module.classifier.r16s.qiime">
<!--   -->
</a>
<h3>Uses of <a href="../../bioLockJ/Module.html" title="class in bioLockJ">Module</a> in <a href="../../bioLockJ/module/classifier/r16s/qiime/package-summary.html">bioLockJ.module.classifier.r16s.qiime</a></h3>
<table class="useSummary" border="0" cellpadding="3" cellspacing="0" summary="Use table, listing subclasses, and an explanation">
<caption><span>Subclasses of <a href="../../bioLockJ/Module.html" title="class in bioLockJ">Module</a> in <a href="../../bioLockJ/module/classifier/r16s/qiime/package-summary.html">bioLockJ.module.classifier.r16s.qiime</a></span><span class="tabEnd">&nbsp;</span></caption>
<tr>
<th class="colFirst" scope="col">Modifier and Type</th>
<th class="colLast" scope="col">Class and Description</th>
</tr>
<tbody>
<tr class="altColor">
<td class="colFirst"><code>class&nbsp;</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../bioLockJ/module/classifier/r16s/qiime/ClosedRefClassifier.html" title="class in bioLockJ.module.classifier.r16s.qiime">ClosedRefClassifier</a></span></code>
<div class="block">This class separates the fasta files into batches of a size defined in the prop file.</div>
</td>
</tr>
<tr class="rowColor">
<td class="colFirst"><code>class&nbsp;</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../bioLockJ/module/classifier/r16s/qiime/DeNovoClassifier.html" title="class in bioLockJ.module.classifier.r16s.qiime">DeNovoClassifier</a></span></code>
<div class="block">Used de novo OTU picking in which all samples must be processed
 as a single batch.</div>
</td>
</tr>
<tr class="altColor">
<td class="colFirst"><code>class&nbsp;</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../bioLockJ/module/classifier/r16s/qiime/MergeOtuTables.html" title="class in bioLockJ.module.classifier.r16s.qiime">MergeOtuTables</a></span></code>&nbsp;</td>
</tr>
<tr class="rowColor">
<td class="colFirst"><code>class&nbsp;</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../bioLockJ/module/classifier/r16s/qiime/OpenRefClassifier.html" title="class in bioLockJ.module.classifier.r16s.qiime">OpenRefClassifier</a></span></code>
<div class="block">Used for Open Ref OTU picking in which all samples must be processed
 as a single batch.</div>
</td>
</tr>
<tr class="altColor">
<td class="colFirst"><code>class&nbsp;</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../bioLockJ/module/classifier/r16s/qiime/QiimePreprocessor.html" title="class in bioLockJ.module.classifier.r16s.qiime">QiimePreprocessor</a></span></code>
<div class="block">This class prepares QIIME input files.</div>
</td>
</tr>
</tbody>
</table>
</li>
<li class="blockList"><a name="bioLockJ.module.classifier.wgs">
<!--   -->
</a>
<h3>Uses of <a href="../../bioLockJ/Module.html" title="class in bioLockJ">Module</a> in <a href="../../bioLockJ/module/classifier/wgs/package-summary.html">bioLockJ.module.classifier.wgs</a></h3>
<table class="useSummary" border="0" cellpadding="3" cellspacing="0" summary="Use table, listing subclasses, and an explanation">
<caption><span>Subclasses of <a href="../../bioLockJ/Module.html" title="class in bioLockJ">Module</a> in <a href="../../bioLockJ/module/classifier/wgs/package-summary.html">bioLockJ.module.classifier.wgs</a></span><span class="tabEnd">&nbsp;</span></caption>
<tr>
<th class="colFirst" scope="col">Modifier and Type</th>
<th class="colLast" scope="col">Class and Description</th>
</tr>
<tbody>
<tr class="altColor">
<td class="colFirst"><code>class&nbsp;</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../bioLockJ/module/classifier/wgs/KrakenClassifier.html" title="class in bioLockJ.module.classifier.wgs">KrakenClassifier</a></span></code>
<div class="block">This class builds the Kraken classifier scripts.</div>
</td>
</tr>
<tr class="rowColor">
<td class="colFirst"><code>class&nbsp;</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../bioLockJ/module/classifier/wgs/MetaphlanClassifier.html" title="class in bioLockJ.module.classifier.wgs">MetaphlanClassifier</a></span></code>
<div class="block">This class builds the scripts used to call Metaphlan for classification of WGS data.</div>
</td>
</tr>
<tr class="altColor">
<td class="colFirst"><code>class&nbsp;</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../bioLockJ/module/classifier/wgs/SlimmClassifier.html" title="class in bioLockJ.module.classifier.wgs">SlimmClassifier</a></span></code>
<div class="block">This class builds the scripts used to call SLIMM for classification of WGS data.</div>
</td>
</tr>
</tbody>
</table>
</li>
<li class="blockList"><a name="bioLockJ.module.parser">
<!--   -->
</a>
<h3>Uses of <a href="../../bioLockJ/Module.html" title="class in bioLockJ">Module</a> in <a href="../../bioLockJ/module/parser/package-summary.html">bioLockJ.module.parser</a></h3>
<table class="useSummary" border="0" cellpadding="3" cellspacing="0" summary="Use table, listing subclasses, and an explanation">
<caption><span>Subclasses of <a href="../../bioLockJ/Module.html" title="class in bioLockJ">Module</a> in <a href="../../bioLockJ/module/parser/package-summary.html">bioLockJ.module.parser</a></span><span class="tabEnd">&nbsp;</span></caption>
<tr>
<th class="colFirst" scope="col">Modifier and Type</th>
<th class="colLast" scope="col">Class and Description</th>
</tr>
<tbody>
<tr class="altColor">
<td class="colFirst"><code>class&nbsp;</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../bioLockJ/module/parser/ParserModule.html" title="class in bioLockJ.module.parser">ParserModule</a></span></code>
<div class="block">The ParserModule is used output standardized tables for any classifier.</div>
</td>
</tr>
</tbody>
</table>
</li>
<li class="blockList"><a name="bioLockJ.module.parser.r16s">
<!--   -->
</a>
<h3>Uses of <a href="../../bioLockJ/Module.html" title="class in bioLockJ">Module</a> in <a href="../../bioLockJ/module/parser/r16s/package-summary.html">bioLockJ.module.parser.r16s</a></h3>
<table class="useSummary" border="0" cellpadding="3" cellspacing="0" summary="Use table, listing subclasses, and an explanation">
<caption><span>Subclasses of <a href="../../bioLockJ/Module.html" title="class in bioLockJ">Module</a> in <a href="../../bioLockJ/module/parser/r16s/package-summary.html">bioLockJ.module.parser.r16s</a></span><span class="tabEnd">&nbsp;</span></caption>
<tr>
<th class="colFirst" scope="col">Modifier and Type</th>
<th class="colLast" scope="col">Class and Description</th>
</tr>
<tbody>
<tr class="altColor">
<td class="colFirst"><code>class&nbsp;</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../bioLockJ/module/parser/r16s/QiimeParser.html" title="class in bioLockJ.module.parser.r16s">QiimeParser</a></span></code>
<div class="block">To see file format: > head otu_table_L2.txt

 # Constructed from biom file #OTU ID 3A.1 6A.1 120A.1 7A.1
 k__Bacteria;p__Actinobacteria 419.0 26.0 90.0 70.0</div>
</td>
</tr>
<tr class="rowColor">
<td class="colFirst"><code>class&nbsp;</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../bioLockJ/module/parser/r16s/RdpParser.html" title="class in bioLockJ.module.parser.r16s">RdpParser</a></span></code>
<div class="block">To see file format: > head 7A_1_reported.tsv

 FCABK7W:1:2105:21787:12788#/1 Root rootrank 1.0 Bacteria domain 1.0
 Firmicutes phylum 1.0 Clostridia class 1.0 Clostridiales order 1.0
 Ruminococcaceae family 1.0 Faecalibacterium genus 1.0</div>
</td>
</tr>
</tbody>
</table>
</li>
<li class="blockList"><a name="bioLockJ.module.parser.wgs">
<!--   -->
</a>
<h3>Uses of <a href="../../bioLockJ/Module.html" title="class in bioLockJ">Module</a> in <a href="../../bioLockJ/module/parser/wgs/package-summary.html">bioLockJ.module.parser.wgs</a></h3>
<table class="useSummary" border="0" cellpadding="3" cellspacing="0" summary="Use table, listing subclasses, and an explanation">
<caption><span>Subclasses of <a href="../../bioLockJ/Module.html" title="class in bioLockJ">Module</a> in <a href="../../bioLockJ/module/parser/wgs/package-summary.html">bioLockJ.module.parser.wgs</a></span><span class="tabEnd">&nbsp;</span></caption>
<tr>
<th class="colFirst" scope="col">Modifier and Type</th>
<th class="colLast" scope="col">Class and Description</th>
</tr>
<tbody>
<tr class="altColor">
<td class="colFirst"><code>class&nbsp;</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../bioLockJ/module/parser/wgs/KrakenParser.html" title="class in bioLockJ.module.parser.wgs">KrakenParser</a></span></code>
<div class="block">To see file format: > head 7A_1_reported.tsv

 FCC6MMAACXX:8:1101:1968:2100#GTATTCTC/1
 d__Bacteria|p__Bacteroidetes|c__Bacteroidia|o__Bacteroidales|f__Bacteroidaceae|g__Bacteroides|s__Bacteroides_vulgatus</div>
</td>
</tr>
<tr class="rowColor">
<td class="colFirst"><code>class&nbsp;</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../bioLockJ/module/parser/wgs/MetaphlanParser.html" title="class in bioLockJ.module.parser.wgs">MetaphlanParser</a></span></code>
<div class="block">To see file format: > head 7A_1_processed.txt

 #SampleID Metaphlan2_Analysis
 #clade_name relative_abundance coverage average_genome_length_in_the_clade estimated_number_of_reads_from_the_clade
 k__Bacteria|p__Bacteroidetes 14.68863 0.137144143537 4234739 580770</div>
</td>
</tr>
<tr class="altColor">
<td class="colFirst"><code>class&nbsp;</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../bioLockJ/module/parser/wgs/SlimmParser.html" title="class in bioLockJ.module.parser.wgs">SlimmParser</a></span></code>
<div class="block">To see file format: > head 7A_1_phylum_reported.tsv

 No.</div>
</td>
</tr>
</tbody>
</table>
</li>
<li class="blockList"><a name="bioLockJ.module.postProcessor">
<!--   -->
</a>
<h3>Uses of <a href="../../bioLockJ/Module.html" title="class in bioLockJ">Module</a> in <a href="../../bioLockJ/module/postProcessor/package-summary.html">bioLockJ.module.postProcessor</a></h3>
<table class="useSummary" border="0" cellpadding="3" cellspacing="0" summary="Use table, listing subclasses, and an explanation">
<caption><span>Subclasses of <a href="../../bioLockJ/Module.html" title="class in bioLockJ">Module</a> in <a href="../../bioLockJ/module/postProcessor/package-summary.html">bioLockJ.module.postProcessor</a></span><span class="tabEnd">&nbsp;</span></caption>
<tr>
<th class="colFirst" scope="col">Modifier and Type</th>
<th class="colLast" scope="col">Class and Description</th>
</tr>
<tbody>
<tr class="altColor">
<td class="colFirst"><code>class&nbsp;</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../bioLockJ/module/postProcessor/RScriptModule.html" title="class in bioLockJ.module.postProcessor">RScriptModule</a></span></code>
<div class="block">This class builds the R script.</div>
</td>
</tr>
</tbody>
</table>
</li>
<li class="blockList"><a name="bioLockJ.module.preProcessor">
<!--   -->
</a>
<h3>Uses of <a href="../../bioLockJ/Module.html" title="class in bioLockJ">Module</a> in <a href="../../bioLockJ/module/preProcessor/package-summary.html">bioLockJ.module.preProcessor</a></h3>
<table class="useSummary" border="0" cellpadding="3" cellspacing="0" summary="Use table, listing subclasses, and an explanation">
<caption><span>Subclasses of <a href="../../bioLockJ/Module.html" title="class in bioLockJ">Module</a> in <a href="../../bioLockJ/module/preProcessor/package-summary.html">bioLockJ.module.preProcessor</a></span><span class="tabEnd">&nbsp;</span></caption>
<tr>
<th class="colFirst" scope="col">Modifier and Type</th>
<th class="colLast" scope="col">Class and Description</th>
</tr>
<tbody>
<tr class="altColor">
<td class="colFirst"><code>class&nbsp;</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../bioLockJ/module/preProcessor/MergePairedReads.html" title="class in bioLockJ.module.preProcessor">MergePairedReads</a></span></code>
<div class="block">This class will merge forward & reverse fastQ files.</div>
</td>
</tr>
<tr class="rowColor">
<td class="colFirst"><code>class&nbsp;</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../bioLockJ/module/preProcessor/TrimPrimers.html" title="class in bioLockJ.module.preProcessor">TrimPrimers</a></span></code>
<div class="block">This utility trims primers configured using regular expressions.</div>
</td>
</tr>
</tbody>
</table>
</li>
<li class="blockList"><a name="bioLockJ.util">
<!--   -->
</a>
<h3>Uses of <a href="../../bioLockJ/Module.html" title="class in bioLockJ">Module</a> in <a href="../../bioLockJ/util/package-summary.html">bioLockJ.util</a></h3>
<table class="useSummary" border="0" cellpadding="3" cellspacing="0" summary="Use table, listing methods, and an explanation">
<caption><span>Methods in <a href="../../bioLockJ/util/package-summary.html">bioLockJ.util</a> with parameters of type <a href="../../bioLockJ/Module.html" title="class in bioLockJ">Module</a></span><span class="tabEnd">&nbsp;</span></caption>
<tr>
<th class="colFirst" scope="col">Modifier and Type</th>
<th class="colLast" scope="col">Method and Description</th>
</tr>
<tbody>
<tr class="altColor">
<td class="colFirst"><code>void</code></td>
<td class="colLast"><span class="typeNameLabel">MetadataUtil.</span><code><span class="memberNameLink"><a href="../../bioLockJ/util/MetadataUtil.html#addColumnToMetadata-bioLockJ.Module-java.lang.String-java.util.Map-java.io.File-">addColumnToMetadata</a></span>(<a href="../../bioLockJ/Module.html" title="class in bioLockJ">Module</a>&nbsp;blje,
                   java.lang.String&nbsp;name,
                   java.util.Map&lt;java.lang.String,java.lang.Integer&gt;&nbsp;map,
                   java.io.File&nbsp;fileDir)</code>
<div class="block">When a new column is added to metadata, this method will add the column, with all row values.</div>
</td>
</tr>
<tr class="rowColor">
<td class="colFirst"><code>static void</code></td>
<td class="colLast"><span class="typeNameLabel">BashScriptUtil.</span><code><span class="memberNameLink"><a href="../../bioLockJ/util/BashScriptUtil.html#addModules-java.io.BufferedWriter-bioLockJ.Module-">addModules</a></span>(java.io.BufferedWriter&nbsp;writer,
          <a href="../../bioLockJ/Module.html" title="class in bioLockJ">Module</a>&nbsp;blje)</code>
<div class="block">Adds cluster modules to script.</div>
</td>
</tr>
<tr class="altColor">
<td class="colFirst"><code>static void</code></td>
<td class="colLast"><span class="typeNameLabel">BashScriptUtil.</span><code><span class="memberNameLink"><a href="../../bioLockJ/util/BashScriptUtil.html#buildScripts-bioLockJ.Module-java.util.List-java.util.List-">buildScripts</a></span>(<a href="../../bioLockJ/Module.html" title="class in bioLockJ">Module</a>&nbsp;blje,
            java.util.List&lt;java.util.List&lt;java.lang.String&gt;&gt;&nbsp;data,
            java.util.List&lt;java.io.File&gt;&nbsp;files)</code>
<div class="block">Scripts are build for the Module based on the data lines provided.</div>
</td>
</tr>
<tr class="rowColor">
<td class="colFirst"><code>static void</code></td>
<td class="colLast"><span class="typeNameLabel">BashScriptUtil.</span><code><span class="memberNameLink"><a href="../../bioLockJ/util/BashScriptUtil.html#buildScripts-bioLockJ.Module-java.util.List-java.util.List-int-">buildScripts</a></span>(<a href="../../bioLockJ/Module.html" title="class in bioLockJ">Module</a>&nbsp;blje,
            java.util.List&lt;java.util.List&lt;java.lang.String&gt;&gt;&nbsp;data,
            java.util.List&lt;java.io.File&gt;&nbsp;files,
            int&nbsp;size)</code>&nbsp;</td>
</tr>
<tr class="altColor">
<td class="colFirst"><code>static void</code></td>
<td class="colLast"><span class="typeNameLabel">BashScriptUtil.</span><code><span class="memberNameLink"><a href="../../bioLockJ/util/BashScriptUtil.html#buildScripts-bioLockJ.Module-java.util.List-java.lang.String-">buildScripts</a></span>(<a href="../../bioLockJ/Module.html" title="class in bioLockJ">Module</a>&nbsp;blje,
            java.util.List&lt;java.util.List&lt;java.lang.String&gt;&gt;&nbsp;data,
            java.lang.String&nbsp;msg)</code>
<div class="block">Some BioLockJExecutors do not have a line for each file, these pass a single fail message.</div>
</td>
</tr>
<tr class="rowColor">
<td class="colFirst"><code>static java.io.File</code></td>
<td class="colLast"><span class="typeNameLabel">BashScriptUtil.</span><code><span class="memberNameLink"><a href="../../bioLockJ/util/BashScriptUtil.html#createMainScript-bioLockJ.Module-">createMainScript</a></span>(<a href="../../bioLockJ/Module.html" title="class in bioLockJ">Module</a>&nbsp;blje)</code>
<div class="block">Create the main script for the Module.</div>
</td>
</tr>
</tbody>
</table>
</li>
</ul>
</li>
</ul>
</div>
<!-- ======= START OF BOTTOM NAVBAR ====== -->
<div class="bottomNav"><a name="navbar.bottom">
<!--   -->
</a>
<div class="skipNav"><a href="#skip.navbar.bottom" title="Skip navigation links">Skip navigation links</a></div>
<a name="navbar.bottom.firstrow">
<!--   -->
</a>
<ul class="navList" title="Navigation">
<li><a href="../../overview-summary.html">Overview</a></li>
<li><a href="../package-summary.html">Package</a></li>
<li><a href="../../bioLockJ/Module.html" title="class in bioLockJ">Class</a></li>
<li class="navBarCell1Rev">Use</li>
<li><a href="../package-tree.html">Tree</a></li>
<li><a href="../../deprecated-list.html">Deprecated</a></li>
<li><a href="../../index-files/index-1.html">Index</a></li>
<li><a href="../../help-doc.html">Help</a></li>
</ul>
</div>
<div class="subNav">
<ul class="navList">
<li>Prev</li>
<li>Next</li>
</ul>
<ul class="navList">
<li><a href="../../index.html?bioLockJ/class-use/Module.html" target="_top">Frames</a></li>
<li><a href="Module.html" target="_top">No&nbsp;Frames</a></li>
</ul>
<ul class="navList" id="allclasses_navbar_bottom">
<li><a href="../../allclasses-noframe.html">All&nbsp;Classes</a></li>
</ul>
<div>
<script type="text/javascript"><!--
  allClassesLink = document.getElementById("allclasses_navbar_bottom");
  if(window==top) {
    allClassesLink.style.display = "block";
  }
  else {
    allClassesLink.style.display = "none";
  }
  //-->
</script>
</div>
<a name="skip.navbar.bottom">
<!--   -->
</a></div>
<!-- ======== END OF BOTTOM NAVBAR ======= -->
</body>
</html>
